<a href="https://colab.research.google.com/github/animesh-11/AI_ML/blob/main/Statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

You are a data analyst at a growing e-commerce company that regularly tracks daily sales across a weekly  reporting cycle. During one such weekly cycle, a system issue caused the loss of one day’s sales data. However, the average daily sales for the full 7-day period had already been calculated before the error occurred.


To maintain accurate business reporting and ensure complete financial records, you must calculate the missing day’s sales based on the provided average and the available sales figures.


Your task is to define a function restore_missing_sale(data) that accepts a list of daily sales with one missing value and the reported average, and returns the missing sale amount.


Input Format

A tuple containing two inputs in the following order:


sales: A list of known daily sales values (float)
sales_avg: The average daily sales (float) for the complete period including the missing day

Output Format


A single value (float) representing the missing sale amount

Constraints

The computed result is rounded to two decimals

Example Case 1


Input

([1200.0, 1350.0, 1280.0, 1420.0, 1500.0, 1300.0], 1371.43)  



Output


1550.01



Example Case 2


Input


([1000.0, 1100.0, 1050.0, 1200.0, 1150.0, 1080.0], 1100.0)



Output


1120.0

In [1]:
from ast import literal_eval

def restore_missing_sale(data):
    sales = sum(data[0])
    sales_total = data[1] * 7
    return sales_total - sales


# (do not edit)
user_input = input()
parsed_input = literal_eval(user_input)
print(round(restore_missing_sale(parsed_input), 2))

([1000.0, 1100.0, 1050.0, 1200.0, 1150.0, 1080.0], 1100.0)
1120.0


You are a data analyst at a ticketing platform that tracks user interest across a variety of events including concerts, plays, and sports. Understanding which events generate the most interaction is key for effective marketing and resource planning.


Your task is to implement the function high_interest_events(data), which analyses the platform's interaction data and identifies which event(s) received the highest user interest.


Input Format

A list of event names (str) representing user interactions on the platform

Output Format

A list of event name(s) (str) that received the highest user interest, sorted in alphabetical order


Constraints

N/A


Example Case 1

Input

['Concert', 'Concert', 'Play', 'Sports', 'Concert', 'Play']


Output

['Concert']


Example Case 2

Input

['Play', 'Play', 'Concert', 'Sports', 'Concert', 'Sports']

Output

['Concert', 'Play', 'Sports']

In [6]:
from ast import literal_eval

def verify_event_popularity(events):
    event_dict = {}
    for event in events:
        if event in event_dict:
            event_dict[event] += 1
        else:
            event_dict[event] = 1

    max_value = max(event_dict.values())
    max_keys = []
    for key, value in event_dict.items():
        if value == max_value:
            max_keys.append(key)
    max_keys.sort()
    return max_keys


# (do not edit)
print(verify_event_popularity(literal_eval(input())))

['Rock', 'Rock', 'Jazz', 'Jazz', 'Jazz', 'Rock']
['Jazz', 'Rock']


Edumedic, a health-tech company managing remote patient monitoring systems, aims to enhance early detection of abnormal health trends among patients recovering at home. One of the most critical metrics tracked is body temperature, as it can be an early sign of infection, inflammation, or fever relapse.


Your task is to develop an automated method to flag highly deviated temperature readings based on statistical analysis. Using historical temperature data for a group of patients, you are to identify any readings that lie significantly outside the typical temperature range, specifically those that fall beyond two standard deviation units from the average.


Your task is to define a function flag_highly_deviated_temperatures(temperatures) that:


Accepts a tuple of numerical values representing body temperature readings of patients
Identifies and returns a list of temperature readings that are highly deviated — those that lie outside the range of two standard deviations from the mean temperature readings

Input Format

A tuple of values (float) representing temperature readings

Output Format

A list of temperature values (float) that are highly deviated from the average body temperature

Constraints

N/A

Example Case 1


Input


(98.6, 98.7, 98.8, 98.9, 99.0, 98.4, 98.5, 98.3, 98.6, 98.7)


Output


[]



Example Case 2


Input


(98.6, 99.1, 97.5, 102.2, 98.7, 98.5, 98.9, 99.0, 98.8, 98.6)


Output


[102.2]



Calculate the mean of the input temperatures tuple.

Calculate the standard deviation of the input temperatures tuple.

Define the lower bound and upper bound of the acceptable range.

Loop through each temperature in the input tuple.

Inside the loop, check if the current temperature is outside of the acceptable range (i.e., it's less than the lower bound OR greater than the upper bound).

If it is, add it to a list that we'll return at the end.

In [8]:
from ast import literal_eval
# removed math library

def flag_highly_deviated_temperatures(temperatures):
    # Code here
    mean_temp = sum(temperatures) / len(temperatures)
    # Calculate standard deviation without math.sqrt
    std_dev = (sum([(x - mean_temp) ** 2 for x in temperatures]) / len(temperatures))**0.5

    lower_bound = mean_temp - 2 * std_dev
    upper_bound = mean_temp + 2 * std_dev

    deviated_temps = []
    for temp in temperatures:
        if temp < lower_bound or temp > upper_bound:
            deviated_temps.append(temp)

    return deviated_temps


# Input and output processing (do not edit)
print(flag_highly_deviated_temperatures(literal_eval(input())))

(98.6, 99.1, 97.5, 102.2, 98.7, 98.5, 98.9, 99.0, 98.8, 98.6)
[102.2]


You work as a data analyst at Edurise, an ed-tech company offering competitive exam coaching. The academic team plans to provide early access to advanced mentoring and problem-solving resources to a very small group of students who scored higher than at least 95 percent of their peers in the batch.


Your task is to define a function top_percentile_students(data) that accepts the student IDs and associated scores, and identifies and returns the IDs of the best performing students.


Input Format


A tuple of tuples, where each inner tuple contains:


student_id: A value (int) representing the unique identifier assigned to each student in the top-ranker batch

score: A number (int or float) indicating the mark secured by the student


Output Format


A list of values (int), each representing the student_id of a student whose score is greater than or equal to at least 95% of the batch

The list must be sorted in ascending order
Your percentile calculation must perform interpolation

Constraints


The input tuple contains at least two entries

Example Case 1

Input

((101, 88.5), (102, 91.0), (103, 76.5), (104, 95.0), (105, 89.0), (106, 78.5),(107, 82.0), (108, 85.5), (109, 92.0), (110, 87.0), (111, 80.0), (112, 84.5),(113, 83.0), (114, 90.0), (115, 86.0), (116, 79.5), (117, 94.0), (118, 88.0), (119, 77.5), (120, 81.0), (121, 86.5), (122, 93.5), (123, 89.5), (124, 90.5),(125, 92.5), (126, 87.5), (127, 85.0), (128, 84.0), (129, 82.5), (130, 91.5))

Output

[104, 117]


Example Case 2

Input


((1, 298), (2, 297), (3, 297), (4, 296), (5, 295),(6, 295), (7, 294), (8, 294), (9, 293), (10, 292))

Output


[1]

In [1]:
from ast import literal_eval
# removed math library

def top_percentile_students(data):
    # Sort data by score
    sorted_data = sorted(data, key=lambda item: item[1])
    scores = [item[1] for item in sorted_data]
    n = len(scores)
    percentile_rank = 0.95 * (n - 1)

    # Calculate 95th percentile score with interpolation
    if percentile_rank.is_integer():
        percentile_score = scores[int(percentile_rank)]
    else:
        lower_index = int(percentile_rank) # Using int() for floor
        upper_index = int(percentile_rank) + 1 # Using int() + 1 for ceiling
        lower_score = scores[lower_index]
        upper_score = scores[upper_index]
        percentile_score = lower_score + (percentile_rank - lower_index) * (upper_score - lower_score)

    # Identify student IDs with scores >= 95th percentile score
    top_students = [student_id for student_id, score in data if score >= percentile_score]

    # Sort the result in ascending order
    top_students.sort()

    return top_students

# Input and output processing (do not edit)
print(top_percentile_students(literal_eval(input())))

((101, 88.5), (102, 91.0), (103, 76.5), (104, 95.0), (105, 89.0), (106, 78.5),(107, 82.0), (108, 85.5), (109, 92.0), (110, 87.0), (111, 80.0), (112, 84.5),(113, 83.0), (114, 90.0), (115, 86.0), (116, 79.5), (117, 94.0), (118, 88.0), (119, 77.5), (120, 81.0), (121, 86.5), (122, 93.5), (123, 89.5), (124, 90.5),(125, 92.5), (126, 87.5), (127, 85.0), (128, 84.0), (129, 82.5), (130, 91.5))
[104, 117]


You are part of the pricing strategy team at a national food chain that operates through independently managed franchise outlets. Each outlet sets its own prices for a common menu, leading to variation across locations.


To improve pricing standardisation and ensure a consistent customer experience, your team wants to identify outlets whose pricing behaviour deviates from the most common pattern observed across the network.


Each outlet reports a list of prices for its local menu. A pricing pattern is defined by two factors:


The average menu price, and
The standard deviation of those prices (i.e., price consistency)

The most common pattern across all outlets is treated as the baseline. Any outlet that does not follow this common pricing pattern should be flagged for review.


Write a function identify_non_standard_outlets(outlet_data) that:


Identifies the most common pricing pattern based on rounded average and standard deviation
Returns the names of all outlets that do not follow this pattern, in alphabetical order
If all outlets follow the same pattern, return the string 'All outlets aligned'


Input Format

A tuple of tuples, where each inner tuple contains:

outlet_name: a string (str) representing the name of the outlet
menu_prices: a list of values (float) representing the prices of items at that outlet

Output Format

A list of outlet names (str) that do not match the common pricing pattern
Or a string 'All outlets aligned' if no such outlets exist

Constraints

The number of outlets is at least 3
Each outlet has at least 3 price values
All prices are non-negative floats
Average and standard deviation must be rounded down to the nearest integer before comparison
An outlet is considered non-standard if its rounded average and standard deviation do not match the most common combination

Example Case 1


Input


(("OutletA", [100, 200, 300]),  ("OutletB", [150, 200, 250]),  ("OutletC", [100, 200, 300]),  ("OutletD", [200, 300, 400]),  ("OutletE", [100, 200, 300]),  ("OutletF", [150, 200, 250]))


Output


['OutletB', 'OutletD', 'OutletF']



Example Case 2


Input


(("OutletA", [10, 20, 30, 40]),  ("OutletB", [10, 20, 30, 40]),  ("OutletC", [10, 20, 30, 40]),  ("OutletD", [10, 20, 30, 40]),  ("OutletE", [10, 20, 30, 40]),  ("OutletF", [10, 20, 30, 40]))


Output


All outlets aligned

In [ ]:
from ast import literal_eval
# removed collections library

def identify_non_standard_outlets(outlet_data):
    patterns = []
    for outlet_name, menu_prices in outlet_data:
        mean_price = sum(menu_prices) / len(menu_prices)
        # Calculate standard deviation
        std_dev = (sum([(x - mean_price) ** 2 for x in menu_prices]) / len(menu_prices))**0.5

        # Round down to the nearest integer
        rounded_mean = int(mean_price)
        rounded_std_dev = int(std_dev)

        patterns.append(((rounded_mean, rounded_std_dev), outlet_name))

    # Find the most common pattern without using collections.Counter
    pattern_counts = {}
    for pattern, _ in patterns:
        if pattern in pattern_counts:
            pattern_counts[pattern] += 1
        else:
            pattern_counts[pattern] = 1

    most_common_pattern = None
    max_count = 0
    for pattern, count in pattern_counts.items():
        if count > max_count:
            max_count = count
            most_common_pattern = pattern

    # Identify outlets that don't follow the most common pattern
    non_standard_outlets = [outlet_name for pattern, outlet_name in patterns if pattern != most_common_pattern]

    # Sort the result in alphabetical order
    non_standard_outlets.sort()

    if not non_standard_outlets:
        return "All outlets aligned"
    else:
        return non_standard_outlets


# Input and output processing (do not edit)
print(identify_non_standard_outlets(literal_eval(input())))

You are a data analyst at an e-commerce platform that revises its product pricing and promotional strategies every two weeks. To make informed decisions, the company wishes to understand how product prices influence customer engagement on the platform.


Your goal is to examine whether there is a strong, weak, or neutral relationship between product price and average time spent by the customer on the corresponding product page.


Your task is to define a function analyze_price_time_relation(data) that accepts product prices and average time spent by customers on each corresponding product pages, calculates the correlation between product prices and time spent, and returns one of the following based on the correlation:


'Strong' – if the absolute correlation is ≥ 0.7
'Weak'– if the absolute correlation is between 0.3 and 0.7
'Neutral'– if the absolute correlation is < 0.3

Input Format


A tuple of two tuples:


product_prices: The first tuple includes product prices (float)

time_spent: The second tuple includes the average time (float) spent by customers on each respective product page


Output Format


A single value (str) indicating the strength of the relationship between product price and time spent

Constraints

N/A


Example Case 1  


Input


((100, 150, 200, 250, 300, 350, 400, 450, 500, 550, 600, 650, 700, 750), (2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5))



Output

Strong



Example Case 2

Input

((100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150, 155, 160, 165),    (5.0, 5.1, 5.2, 5.1, 5.0, 5.1, 5.2, 5.0, 5.1, 5.2, 5.0, 5.1, 5.2, 5.0))


Output

Neutral

In [2]:
def analyze_price_time_relation(data):
    """
    Analyzes the correlation between product prices and average time spent on product pages.

    Args:
        data: A tuple of two tuples. The first contains product prices (floats)
              and the second contains average time spent (floats).

    Returns:
        A string indicating the strength of the relationship ('Strong', 'Weak', or 'Neutral').
    """
    # Unpack the two tuples from the input data
    prices = data[0]
    times_spent = data[1]

    # Calculate the mean of each list.
    mean_prices = sum(prices) / len(prices)
    mean_times = sum(times_spent) / len(times_spent)

    # Calculate the numerator of the Pearson correlation formula (Covariance).
    # This is the sum of the products of the deviations from the mean.
    numerator = sum((prices[i] - mean_prices) * (times_spent[i] - mean_times) for i in range(len(prices)))

    # Calculate the two components of the denominator (product of standard deviations).
    # This involves the sum of the squared deviations from the mean for each list.
    price_deviations_squared = sum((p - mean_prices) ** 2 for p in prices)
    time_deviations_squared = sum((t - mean_times) ** 2 for t in times_spent)

    # Check for an edge case where there is no variance in one of the data sets.
    # This would cause a division by zero. In this case, correlation is undefined.
    if price_deviations_squared == 0 or time_deviations_squared == 0:
        return 'Neutral'

    # Calculate the denominator (square root of the product of the squared deviations).
    denominator = (price_deviations_squared * time_deviations_squared) ** 0.5

    # Calculate the final correlation coefficient.
    correlation = numerator / denominator

    # Determine the strength of the relationship based on the absolute correlation.
    absolute_correlation = abs(correlation)

    if absolute_correlation >= 0.7:
        return 'Strong'
    elif 0.3 <= absolute_correlation < 0.7:
        return 'Weak'
    else:  # absolute_correlation < 0.3
        return 'Neutral'

# Example Case 1
case1_input = ((100, 150, 200, 250, 300, 350, 400, 450, 500, 550, 600, 650, 700, 750),
               (2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5))
case1_output = analyze_price_time_relation(case1_input)
print(f"Example Case 1 Output: {case1_output}")

# Example Case 2
case2_input = ((100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150, 155, 160, 165),
               (5.0, 5.1, 5.2, 5.1, 5.0, 5.1, 5.2, 5.0, 5.1, 5.2, 5.0, 5.1, 5.2, 5.0))
case2_output = analyze_price_time_relation(case2_input)
print(f"Example Case 2 Output: {case2_output}")


Example Case 1 Output: Strong
Example Case 2 Output: Neutral


You are a quantitative analyst at a major investment firm. Your primary role involves assessing the risk associated with various financial portfolios. A key aspect of risk analysis is understanding the volatility of returns, which is quantified by measures of variability. The higher the variability, the higher the perceived risk.


You are given a series of daily percentage returns for a specific investment portfolio over a period. Your task is to calculate two critical risk metrics: the Population Variance and the Population Standard Deviation of these daily returns. Additionally, you need to determine if the portfolio's volatility exceeds predefined acceptable thresholds for both metrics.


Input Format

A tuple containing three elements:


A list of daily percentage returns (float or int) for the portfolio

A positive float specifying the acceptable variance threshold

A positive float specifying the acceptable standard deviation threshold


Output Format

The function returns a string describing the portfolio's risk level:


'High Risk: Exceeds Both Thresholds' – if both variance and standard deviation exceed their thresholds
'Moderate Risk: High Variance Only' – if only variance exceeds its threshold
'Moderate Risk: High Standard Deviation Only' – if only standard deviation exceeds its threshold
'Low Risk: Within Acceptable Limits' – if neither exceeds the

Constraints

The daily_returns list contains at least 2 return value
If the list contains only 1 return value, the variance and standard deviation are treated as 0, and the portfolio will be classified as 'Low Risk: Within Acceptable Limits'.

Example Case 1

Input

([0.02, -0.01, 0.03, 0.05, -0.02, 0.01], 0.0005, 0.02)


Output


High Risk: Exceeds Both Thresholds


Example Case 2

Input

([0.001, 0.002, 0.0015, 0.0025, 0.001], 0.0001, 0.01)


Output

Low Risk: Within Acceptable Limits

In [ ]:
from ast import literal_eval

def analyze_portfolio_risk(data):
    """
    Analyzes the risk of a portfolio based on its daily returns,
    population variance, and population standard deviation.

    Args:
        data: A tuple containing:
            - daily_returns: A list of daily percentage returns (float or int).
            - variance_threshold: The acceptable variance threshold (float).
            - std_dev_threshold: The acceptable standard deviation threshold (float).

    Returns:
        A string describing the portfolio's risk level.
    """
    daily_returns, variance_threshold, std_dev_threshold = data

    if len(daily_returns) <= 1:
        return 'Low Risk: Within Acceptable Limits'

    mean_return = sum(daily_returns) / len(daily_returns)

    # Calculate Population Variance
    population_variance = sum([(x - mean_return) ** 2 for x in daily_returns]) / len(daily_returns)

    # Calculate Population Standard Deviation
    population_std_dev = population_variance ** 0.5

    exceeds_variance = population_variance > variance_threshold
    exceeds_std_dev = population_std_dev > std_dev_threshold

    if exceeds_variance and exceeds_std_dev:
        return 'High Risk: Exceeds Both Thresholds'
    elif exceeds_variance:
        return 'Moderate Risk: High Variance Only'
    elif exceeds_std_dev:
        return 'Moderate Risk: High Standard Deviation Only'
    else:
        return 'Low Risk: Within Acceptable Limits'

# Input and output processing (do not edit)
user_input = literal_eval(input())
print(analyze_portfolio_risk(user_input))

You work as a data analyst at Edurise, an ed-tech company offering competitive exam coaching. The product team is aiming to enhance personalised learning recommendations by studying weekly student engagement patterns. They are particularly interested in understanding if students who invest more time on the platform are also more likely to engage actively with practice questions.


Your task is to define a function compute_covariance(data) that accepts the time spent on the platform by students as well as the questions attempted by students, and computes and returns the covariance between these two metrics


Input Format


A tuple of two tuples, each containing 7 numerical values:


The first tuple represents the time spent (in minutes) (int) on the platform each day in a week

The second tuple represents the number of questions (int) attempted each day in a week


Output Format


A single value (float) representing the covariance between time spent on the platform and number of questions attempted

Constraints


The final result is rounded to 2 decimal places


Example Case 1

Input

((45, 59, 40, 35, 66, 79, 72), (29, 22, 33, 43, 48, 42, 50))


Output

58.49


Example Case 2

Input

((30, 35, 40, 38, 42, 45, 50), (60, 55, 50, 52, 48, 45, 40))


Output

-36.86

In [ ]:
from ast import literal_eval

def compute_covariance(data):
    """
    Computes the covariance between two sets of data.

    Args:
        data: A tuple of two tuples, each containing numerical values.
              The first tuple represents time spent, the second represents questions attempted.

    Returns:
        A float representing the covariance, rounded to 2 decimal places.
    """
    time_spent = data[0]
    questions_attempted = data[1]

    n = len(time_spent)
    mean_time = sum(time_spent) / n
    mean_questions = sum(questions_attempted) / n

    covariance = sum([(time_spent[i] - mean_time) * (questions_attempted[i] - mean_questions) for i in range(n)]) / n

    return round(covariance, 2)

# Input and output processing (do not edit)
print(compute_covariance(literal_eval(input())))

You are working with a support operations team that tracks customer wait times across multiple service regions. The team uses these wait times to measure how well each region complies with its Service Level Agreement (SLA).


According to the SLA, the median wait time in any region should not exceed a specified threshold (in minutes). Your task is to identify all regions that fail to meet this SLA requirement.


Write a function failing_regions(wait_data, sla) that computes the median wait time for each region and returns the names of regions where the median wait time is greater than the SLA threshold.


Input Format


A tuple with two elements:

wait_data: A dictionary where each key is a region name (string) and each value is a list of wait times (float) in minutes
sla: A value (float) representing the SLA threshold

Output Format

A list of region names (str) whose median wait time exceeds the SLA threshold, sorted in alphabetical order
If all regions comply with the SLA, return the string 'All regions meet SLA'

Constraints

All wait time values are non-negative floats
SLA threshold values are non-negative floats
Round the median wait time to two decimal places before comparing it with the SLA threshold


Example Case 1



Input

(

    {

        "North": [5.0, 7.2, 6.8, 8.0, 9.1, 6.5, 7.0],

        "South": [12.5, 13.0, 14.1, 15.2, 12.0, 14.8],

        "East": [6.0, 6.1, 6.2, 6.3, 6.4, 6.5],

        "West": [10.0, 10.5, 9.8, 11.2, 10.3]

    },

    10.0

)


Output

["South", "West"]


Example Case 2

Input

(

    {

        "North": [2.0, 2.5, 3.0, 3.5, 4.0],

        "South": [9.9, 10.0, 10.1, 10.0, 10.2],

        "East": [5.0, 5.5, 6.0, 5.5, 6.0],

        "West": [7.0, 6.5, 7.5, 7.0, 6.8]

    },

    9.5

)


Output

["South"]

In [3]:
from ast import literal_eval

def failing_regions(data):
    """
    Identifies regions where the median wait time exceeds the SLA threshold.

    Args:
        data: A tuple containing:
            - wait_data: A dictionary with region names as keys and lists of wait times as values.
            - sla: The SLA threshold (float).

    Returns:
        A list of region names exceeding the SLA, sorted alphabetically,
        or "All regions meet SLA" if none exceed.
    """
    wait_data, sla = data
    failing = []

    for region, wait_times in wait_data.items():
        sorted_wait_times = sorted(wait_times)
        n = len(sorted_wait_times)
        if n % 2 == 0:
            median = (sorted_wait_times[n // 2 - 1] + sorted_wait_times[n // 2]) / 2
        else:
            median = sorted_wait_times[n // 2]

        rounded_median = round(median, 2)

        if rounded_median > sla:
            failing.append(region)

    failing.sort()

    if not failing:
        return "All regions meet SLA"
    else:
        return failing

# Input and output processing (do not edit)
print(failing_regions(literal_eval(input())))

(     {         "North": [5.0, 7.2, 6.8, 8.0, 9.1, 6.5, 7.0],         "South": [12.5, 13.0, 14.1, 15.2, 12.0, 14.8],         "East": [6.0, 6.1, 6.2, 6.3, 6.4, 6.5],         "West": [10.0, 10.5, 9.8, 11.2, 10.3]     },     10.0 )
['South', 'West']


You are helping a wellness platform analyse data from wearable devices that track heart rate and sleep duration. The goal is to detect users whose heart rate is strongly negatively correlated with their sleep duration — a potential indicator of chronic stress or burnout.


Each user’s wearable data consists of several daily records, where each record logs the average heart rate and corresponding sleep duration for that day.


You have been asked to identify all users who show a strong negative correlation (≤ –0.70) between their heart rate and sleep duration. This correlation must be computed manually, without using any external libraries.


Write a function find_stress_linked_users(user_data) that calculates the Pearson correlation coefficient between heart rate and sleep duration for each user and returns the names of users who show a strong negative correlation (≤ –0.70)

Input Format

A tuple of tuples, where each inner tuple contains:

user_id: a string representing the user’s ID
records: a list of tuples (float, float) representing daily records in the form (heart_rate, sleep_duration)
Output Format

A list of user IDs (str) whose rounded correlation coefficient is ≤ –0.70, sorted in alphabetical order
If no such users exist, return the string 'No stress-linked patterns detected'

Constraints

All values are non-negative floats
Round the final correlation to 2 decimal places before comparison

Example Case 1

Input

(('U1', [(70, 6), (75, 6), (80, 6), (85, 6), (90, 6)]),

 ('U2', [(90, 4), (85, 5), (80, 6), (75, 7), (70, 8)]),

 ('U3', [(60, 7), (65, 7.1), (70, 7.2), (75, 7.3), (80, 7.4)]))


Output

['U2']



Example Case 2


Input


(('U1', [(70, 6), (75, 6.5), (80, 6), (85, 6.5), (90, 6)]),

 ('U2', [(90, 5), (90, 5), (90, 5), (90, 5), (90, 5)]),

 ('U3', [(60, 7), (65, 6.8), (70, 7.1), (75, 6.9), (80, 7)]))


Output


No stress-linked patterns detected

In [4]:
from ast import literal_eval

def find_stress_linked_users(user_data):
    """
    Identifies users with a strong negative correlation between heart rate and sleep duration.

    Args:
        user_data: A tuple of tuples, where each inner tuple contains:
            - user_id: A string representing the user’s ID.
            - records: A list of tuples (float, float) representing daily records (heart_rate, sleep_duration).

    Returns:
        A list of user IDs (str) whose rounded correlation coefficient is ≤ –0.70, sorted alphabetically,
        or "No stress-linked patterns detected" if no such users exist.
    """
    stress_linked_users = []

    for user_id, records in user_data:
        if len(records) < 2:
            continue # Cannot calculate correlation with less than 2 data points

        heart_rates = [record[0] for record in records]
        sleep_durations = [record[1] for record in records]
        n = len(records)

        # Calculate means
        mean_heart_rate = sum(heart_rates) / n
        mean_sleep_duration = sum(sleep_durations) / n

        # Calculate numerator (covariance)
        numerator = sum([(heart_rates[i] - mean_heart_rate) * (sleep_durations[i] - mean_sleep_duration) for i in range(n)])

        # Calculate denominators (product of standard deviations)
        heart_rate_deviations_squared = sum([(hr - mean_heart_rate) ** 2 for hr in heart_rates])
        sleep_duration_deviations_squared = sum([(sd - mean_sleep_duration) ** 2 for sd in sleep_durations])

        denominator = (heart_rate_deviations_squared * sleep_duration_deviations_squared) ** 0.5

        # Calculate correlation coefficient
        if denominator == 0:
            correlation = 0.0 # Handle cases with no variance
        else:
            correlation = numerator / denominator

        # Round correlation to 2 decimal places
        rounded_correlation = round(correlation, 2)

        # Check for strong negative correlation
        if rounded_correlation <= -0.70:
            stress_linked_users.append(user_id)

    # Sort the result in alphabetical order
    stress_linked_users.sort()

    if not stress_linked_users:
        return "No stress-linked patterns detected"
    else:
        return stress_linked_users

# Input and output processing (do not edit)
print(find_stress_linked_users(literal_eval(input())))

(('U1', [(70, 6), (75, 6.5), (80, 6), (85, 6.5), (90, 6)]),  ('U2', [(90, 5), (90, 5), (90, 5), (90, 5), (90, 5)]),  ('U3', [(60, 7), (65, 6.8), (70, 7.1), (75, 6.9), (80, 7)]))
No stress-linked patterns detected
